feat(request_options): add request options to any method using API calls

Jerskouille opened the pull request:
Should be good now!

Please let's not write the associated tests for every method.

redox commented on lib/algolia/client.rb:
```diff
@@ -102,8 +102,9 @@ def with_rate_limits(end_user_ip, rate_limit_api_key, &block)
+    # @param request_options contains extra parameters to send with your query
     #
-    def multiple_queries(queries, index_name_key = :index_name, strategy = "none")
+    def multiple_queries(queries, index_name_key = :index_name, strategy = "none", request_options = {})
```
Prototypes like this one really don't look ruby-ish :/

Jerskouille replied:
Agreed on all of those.
WDYT about:
```ruby
def multiple_queries(queries, options = nil, strategy = nil)
  if options.is_a?(Hash)
    index_name_key = options.delete(:index_name_key) || options.delete('index_name_key') || :index_name
    strategy = options.delete(:strategy) || options.delete('strategy') || 'none'
    request_options = options.delete(:request_options) || options.delete('request_options') || {}
  else
    index_name_key = options || :index_name
    strategy ||= 'none'
    request_options = {}
  end
  # ...
end
```

redox replied:
Yes, that would be 💯

redox commented on lib/algolia/client.rb:
```diff
@@ -185,26 +192,31 @@ def delete_index!(name)
+    # @param request_options contains extra parameters to send with your query
     #
-    def get_logs(offset = 0, length = 10, type = "all")
+    def get_logs(offset = 0, length = 10, type = "all", request_options = {})
```
Same :/

redox commented on lib/algolia/client.rb:
```diff
@@ -233,8 +245,9 @@ def get_api_key(key)
+    #  @param request_options contains extra parameters to send with your query
     #
-    def add_api_key(obj, validity = 0, maxQueriesPerIPPerHour = 0, maxHitsPerQuery = 0, indexes = nil)
+    def add_api_key(obj, validity = 0, maxQueriesPerIPPerHour = 0, maxHitsPerQuery = 0, indexes = nil, request_options = {})
```
Same :/

redox commented on lib/algolia/client.rb:
```diff
@@ -413,16 +429,20 @@ def thread_local_hosts(read)
+      hs = {}
+      extra_headers = request_options[:headers] || request_options['headers'] || {}
+      @headers.each { |key, val| hs[key.to_s] = val }
+      extra_headers.each { |key, val| hs[key.to_s] = val }
```
Why do you explicitly call `key.to_s`? Coud we write `hs = @headers.merge(request_options[:headers] || request_options['headers'] || {})` instead?

Jerskouille replied:
In order to make sure we're correctly handling `headers: { MyCustomHeader: 'xxx' }` instead of `headers: { 'MyCustomHeader' => 'xxx'}`.

redox replied:
But this will be automatically done by the underlying `httpclient` library, no?

Jerskouille replied:
You're right, I had forgotten that ruby guaranteed the iteration on hashes to be in the order of insertion (JS doesn't).

redox commented on lib/algolia/index.rb:
```diff
@@ -224,36 +236,39 @@ def get_object(objectID, attributesToRetrieve = nil)
+    # @param request_options contains extra parameters to send with your query
     #
-    def wait_task(taskID, timeBeforeRetry = 100)
+    def wait_task(taskID, timeBeforeRetry = 100, request_options = {})
```
Same :/

Jerskouille replied:
Here I would leave it like this. We also have `def add_object(obj, objectID = nil, request_options = {})` for instance.

If I do this, I should as well do this for each and every method.

Jerskouille commented:
Since this PR:
- delete_by_query was modified
- we've added query rules endpoints
- we've added delete_by

Jerskouille commented:
I've just rebased on master and fixed the conflicts.

I will review one last time myself and merge with my fingers crossed.
This PR has definitely been opened for too long.
